### DISCUSSION TITLE
React native support

### DISCUSSION THREAD

**Discussion by torver213** [privateOmega/html-to-docx#206] [General]
Is this library supported in react native? If no, how can I make it compatible with react native?

**Comment by gradaom** [comment]
React Native has its own limitations and differences compared to traditional web development. It uses native components instead of HTML and runs JavaScript in a separate context.

To make `html-to-docx` work with React Native, you would need to find an alternative solution or create a custom implementation. Here are a few possible approaches:

1. Use a different library: Look for a React Native-specific library that can convert HTML to DOCX. There may be libraries specifically designed for working with React Native that can handle this task.

2. Utilize a webview: As an alternative, you could use a webview component in React Native to render the HTML content and then use the `html-to-docx` library within the webview to generate the DOCX file. This approach allows you to leverage existing Node.js libraries within a web-based environment.

3. Convert on a server: If you have control over a server-side environment, you can send the HTML content to the server via an API call from your React Native app, and then use the `html-to-docx` library on the server to generate the DOCX file. Your React Native app can then download the resulting DOCX file.

4. Write a custom implementation: Another option is to write your own converter from HTML to DOCX specifically tailored for React Native. You can leverage React Native's built-in APIs and third-party libraries to parse the HTML markup, extract relevant information, and generate a DOCX file using the appropriate file format specifications.

In summary, `html-to-docx` is not natively supported in React Native. You'll need to explore alternative libraries, use a webview, convert on a server, or develop a custom implementation to achieve HTML to DOCX conversion within a React Native environment.

**Reply by torver213** [reply]
@wb-ts thank you for this detailed explanation. I will choose between your suggested options

**Comment by mwaleed-39** [comment]
I faced some issue in react native. when we pressed upload image, the error message shows:
CODE:

import React, { useState } from 'react'
import { Alert, Button, View,StyleSheet } from 'react-native'
import * as ImagePicker from 'react-native-image-picker'
import {launchImageLibrary} from 'react-native-image-picker'

const options ={
  title: 'Select Image',
    type: 'library',
    options: {
      maxHeight:200,
      maxWidth:200,
      selectionLimit: 1,
      mediaType: 'photo',
      includeBase64: false,
    }
}

const Request = () => {
   const opengallery = async()=>{
    const image = await ImagePicker.launchImageLibrary(options);
    console.log(image)
   }

  return (
    <View style = {styles.container}>
       <Button title='Upload Image' onPress={opengallery}/>
    </View>
  )
}
export default Request;

const styles = StyleSheet.create({
  container:{
    flex:1,
    alignItems:'center',
    justifyContent:'center'
  }
})

when we pressed upload button, the error message shows:

Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

I create API:
app.post('/finerequest', (req, res, next) => {

  const file = req.files.image;
  cloudinary.uploader.upload(file.tempFilePath, (err, result) => {
    console.log(result)
    const requestId = req.body.request_id
    const fine = new Fine({
      _id: new mongoose.Types.ObjectId,
      request_id: requestId,
      image: result.url
    })
    fine.save()
      .then(result => {
        res.status(200).json({
          message: "Fine and Vehicle Repair Request added",
          response: result
        })
      })
      .catch(err => {
        res.status(500).json({
          message: "Fine and Vehicle request not added",
          error: err
        })
      })
  })
})
How to call this API, when we upload image in front end side, the image will be stored in mongodb

**Comment by gradaom** [comment]
To call the API and store the uploaded image in MongoDB, you can make some changes to your React Native code and use a library like axios to send the image data to your backend.

First, make sure you have axios installed in your project by running the following command:

```
npm install axios
```

Then, update your React Native code as follows:

```javascript
import React, { useState } from 'react';
import { Alert, Button, View, StyleSheet } from 'react-native';
import * as ImagePicker from 'react-native-image-picker';
import axios from 'axios';

const options = {
  title: 'Select Image',
  type: 'library',
  options: {
    maxHeight: 200,
    maxWidth: 200,
    selectionLimit: 1,
    mediaType: 'photo',
    includeBase64: false,
  },
};

const Request = () => {
  const [imageData, setImageData] = useState(null);

  const opengallery = async () => {
    const image = await ImagePicker.launchImageLibrary(options);
    console.log(image);
    setImageData(image);
  };

  const uploadImage = () => {
    if (imageData) {
      const formData = new FormData();
      formData.append('image', {
        name: imageData.fileName,
        type: imageData.type,
        uri: imageData.uri,
      });

      axios
        .post('http://your-api-endpoint/finerequest', formData)
        .then((response) => {
          console.log(response.data);
          // Handle success response here
        })
        .catch((error) => {
          console.error(error);
          // Handle error response here
        });
    }
  };

  return (
    <View style={styles.container}>
      <Button title="Open Gallery" onPress={opengallery} />
      <Button title="Upload Image" onPress={uploadImage} />
    </View>
  );
};

export default Request;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

In the above code:

- We added a new state variable `imageData` using `useState(null)` to store the selected image data.
- In the `opengallery` function, we updated `setImageData(image)` to set the selected image data in the state.
- The `uploadImage` function is called when the "Upload Image" button is pressed. It checks if `imageData` is not null, creates a `FormData` object, and appends the image details to it. Then, an axios POST request is sent to your API endpoint along with the form data containing the image.
- Replace `'http://your-api-endpoint/finerequest'` with the actual URL of your API endpoint.

On the backend side, you can use Multer middleware or any other library to handle file uploads and save the image to MongoDB. The example you provided seems to use Cloudinary for image uploading, so make sure you have the necessary setup for Cloudinary in your backend code.

Make sure to adjust the implementation on the backend side based on your requirements and the specific libraries you are using.

**Comment by mwaleed-39** [comment]
when we use this code, it shows error message
Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

**Comment by gradaom** [comment]
Please make sure that you installed the `react-native-image-picker` correctly.

**Comment by gradaom** [comment]
And, please confirm the availability of the `launchImageLibrary` function:
   - Check the version of the `react-native-image-picker` package you have installed. Starting from version 4.0.0, the `launchImageLibrary` function is deprecated, so you should use `launchImageLibrary` instead.
   - Make sure you are using the correct function name according to the version of the library you have installed.

**Comment by mwaleed-39** [comment]
I faced some issue in react native. when we pressed upload image, the error
message shows:

How to call this API, when we upload image in front end side, the image
will be stored in mongodb
I create API:
app.post('/finerequest', (req, res, next) => {

const file = req.files.image;
cloudinary.uploader.upload(file.tempFilePath, (err, result) => {
console.log(result)
const requestId = req.body.request_id
const fine = new Fine({
_id: new mongoose.Types.ObjectId,
request_id: requestId,
image: result.url
})
fine.save()
.then(result => {
res.status(200).json({
message: "Fine and Vehicle Repair Request added",
response: result
})
})
.catch(err => {
res.status(500).json({
message: "Fine and Vehicle request not added",
error: err
})
})
})
})

CODE:

import React, { useState } from 'react'
import { Alert, Button, View,StyleSheet } from 'react-native'
import * as ImagePicker from 'react-native-image-picker'
import {launchImageLibrary} from 'react-native-image-picker'

const options ={
title: 'Select Image',
type: 'library',
options: {
maxHeight:200,
maxWidth:200,
selectionLimit: 1,
mediaType: 'photo',
includeBase64: false,
}
}

const Request = () => {
const opengallery = async()=>{
const image = await ImagePicker.launchImageLibrary(options);
console.log(image)
}

return (

)
}
export default Request;

const styles = StyleSheet.create({
container:{
flex:1,
alignItems:'center',
justifyContent:'center'
}
})

when we pressed upload button, the error message shows:

Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

Again and again same error show, we Image picker library, I think it has
been deprecated

**Comment by gradaom** [comment]
``` javascript
import React from 'react';
import { Alert, Button, View, StyleSheet } from 'react-native';
import * as ImagePicker from 'react-native-image-picker';

const options = {
  title: 'Select Image',
  mediaType: 'photo',
  maxHeight: 200,
  maxWidth: 200,
  selectionLimit: 1,
  includeBase64: false,
};

const Request = () => {
  const openGallery = async () => {
    ImagePicker.launchImageLibrary(options, (response) => {
      console.log(response);
    });
  };

  return (
    <View style={styles.container}>
      <Button onPress={openGallery} title="Upload Image" />
    </View>
  );
};

export default Request;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

**Reply by mwaleed-39** [reply]
WARN  Possible Unhandled Promise Rejection (id: 7):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null
    at anonymous (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:170577:43)
    at tryCallTwo (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:61:9)
    at doResolve (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:216:25)
    at Promise (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:82:14)
    at imageLibrary (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:170576:23)
    at launchImageLibrary (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:170536:118)   
    at ?anon_0_ (http://192.168.10.1:19000/Src\Screens\Fine_repair.bundle?platform=android&hot=false&dev=true&minify=false&modulesOnly=true&runModule=false&shallow=true:25:39)
    at next (native)
    at asyncGeneratorStep (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:4641:26)      
    at _next (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:4660:29)
    at anonymous (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:4665:14)
    at tryCallTwo (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:61:9)
    at doResolve (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:216:25)
    at Promise (/Users/kudo/01_Work/Repos/expo/expo/android/versioned-react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/5a4n485h/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:82:14)
    at anonymous (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:4657:25)
    at apply (native)
    at openGallery (http://192.168.10.9:19000/Src\Screens\Fine_repair.bundle?platform=android&hot=false&dev=true&minify=false&modulesOnly=true&runModule=false&shallow=true:30:26)
    at _performTransitionSideEffects (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62245:22)
    at _receiveSignal (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62201:45)
    at onResponderRelease (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62068:34)     
    at apply (native)
    at invokeGuardedCallbackProd (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62653:21)
    at apply (native)
    at invokeGuardedCallback (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62739:42)  
    at apply (native)
    at invokeGuardedCallbackAndCatchFirstError 
(http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62742:36)
    at executeDispatch (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62806:48)        
    at executeDispatchesInOrder (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:62823:26)
    at executeDispatchesAndRelease (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63929:35)
    at executeDispatchesAndReleaseTopLevel (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63936:43)
    at forEach (native)
    at forEachAccumulated (http://192.168.10,1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63265:22)     
    at runEventsInBatch (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63947:27)       
    at runExtractedPluginEventsInBatch (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:64006:25)
    at anonymous (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63987:42)
    at batchedUpdates$1 (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:74577:20)       
    at batchedUpdates (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63918:36)
    at _receiveRootNodeIDEvent (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:63986:23)    at receiveTouches (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:64029:34)
    at apply (native)
    at __callFunction (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:2826:36)
    at anonymous (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:2587:31)
    at __guard (http://192.168.10.1:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:2777:15)
    at callFunctionReturnFlushedQueue (http://192.168.10.9:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false:2586:21)

It give me a error

**Reply by mwaleed-39** [reply]
In Package.json: we use "react-native-image-picker": "^5.6.0",

**Comment by mwaleed-39** [comment]
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, View } from 'react-native';
import React from 'react';
import Reg_form from './Frontend/Components/RegForm';
import Login_form from './Frontend/Components/Signin';
import ForgotPassword from './Frontend/Components/ForgotPassword';
import { ApplicationProvider, Layout, Text } from '@ui-kitten/components';
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import {
  createDrawerNavigator,
  DrawerContentScrollView,
  DrawerItemList,
  DrawerItem,
} from '@react-navigation/drawer';
import * as eva from '@eva-design/eva';
import ChangePassword from './Frontend/Components/ChangePassword';
import VerifyOTP from './Frontend/Components/VerifyOTP';
import Home from './Frontend/Components/Home Screen';
import Profile from './Frontend/Components/Profile';
import Icon from 'react-native-vector-icons'
const Stack = createNativeStackNavigator();
const Drawer = createDrawerNavigator();

const CustomDrawer = props => {

  return (
    <View style={{ flex: 1 }}>
      <DrawerContentScrollView {...props}>
        <View
          style={{
            alignSelf: "center",
            padding: 20,

          }}
        >
          <View>
            <Text style={{ fontSize: 18, fontWeight: "bold" }}> App</Text>
          </View>

        </View>
        <DrawerItemList {...props} />
        <LogoutButton iconName="sign-out" color="black" />
      </DrawerContentScrollView>

    </View>
  );
};



function DrawerRoutes() {


  return (

    <Drawer.Navigator initialRouteName="Home"

      drawerContent={props => <CustomDrawer {...props} />}
    >

      <Drawer.Screen name="Home" component={Home}

        options={{
          headerShown: true,
          
          title: 'Home',
          drawerIcon: ({ focused, size }) => (
            <Icon
              name="home"
              size={size}
              color='black'

            />

          )
        }}
      />
      <Drawer.Screen name="Profile" component={Profile}

        options={{
          headerShown: true,
          title: 'Profile',
          drawerIcon: ({ focused, size }) => (
            <Icon
              name="user"
              size={size}
              color='black'

            />

          )
        }}
      />
    



    </Drawer.Navigator>
  );
}

export default function App() {
  return (
    <ApplicationProvider {...eva} theme={eva.light}> 
    
    <NavigationContainer>
    <Stack.Navigator>
              <Stack.Screen name="Registration" component={Reg_form} />
              <Stack.Screen name="Log In" component={Login_form}/>
              <Stack.Screen name="Forgot" component={ForgotPassword}/>
              <Stack.Screen name="Password" component={ChangePassword}/>
              <Stack.Screen name="Verification" component={VerifyOTP}/>
              <Stack.Screen name="Home" component={Home}  />
              <Stack.Screen name="DrawerRoutes" component={DrawerRoutes}/>
             
  </Stack.Navigator>
  
  </NavigationContainer>
    
  </ApplicationProvider>
  );
}

When we log in to the App, then navigate into the Home Screen. Then we have not visible Drawer navigation
